<h1> Packet Loss <h1>

<h3> 1) Determine the end-to-end transmission duration for a packet. <h3>

In [ ]:
#End to end = prox + tx
#tx = L/C
#tprop= l/v

**Paramters:**

In [3]:
#Propagation delay in ms
t_prop_1 = 10 
t_prop_2 = 10 
t_prop_3 = 1 
t_prop_4 = 10

#Capacity in bits
capacity_1 = 20_000_000
capacity_2 = 4_000_000
capacity_3 = 800_000
capacity_4 = 80_000

#Packetsize in bits
packet_size = 250 * 8

**Calculations:**

1.$t_{tx} = L/C$ <br>
2.$t_E2E_1 = 
\sum_{n=1}^{4} t_{tx_i} +
\sum_{n=1}^{4} t_{prop_i}
$

In [20]:

#Calculate transmissons delay
t_tx_1 = packet_size / capacity_1 * 1000
t_tx_2 = packet_size / capacity_2 * 1000
t_tx_3 = packet_size / capacity_3 * 1000
t_tx_4 = packet_size / capacity_4 * 1000



#Calculate end-to-end transmisson for one packet
t_E2E_1 = 0
for i in range(1,5):
     # Erstelle den Variablennamen mit f-string-Formatierung
    t_prop_i = f"t_prop_{i}"
    t_tx_i = f"t_tx_{i}"
    print("transmission delay: " + str(globals()[t_tx_i]) + "ms")
    
    # Rufe die Variable auf (dies ist nur ein Beispiel, du kannst hier deine eigene Logik hinzufügen)
    t_E2E_1 =+ globals()[t_prop_i]
    t_E2E_1 =+ globals()[t_tx_i]

print("End-to-End-Delay: " + str(t_E2E_1) + "ms")

transmission delay: 0.1ms
transmission delay: 0.5ms
transmission delay: 2.5ms
transmission delay: 25.0ms
End-to-End-Delay: 25.0ms


<H3> 2) How long does it take to transmit 35 packets from Q to Z? Assume that the buffers are large enough so that no packets are lost. <h3>

**Formel**

1. $TE2E(n) = TE2E(1) + (n - 1) * L_{\text{min}}(C_1) = \sqrt{C_1}$






In [18]:
t_E2E_35 = round(t_E2E_1 + ((35-1) * t_tx_4), 3)
print("End-to-End-Delay for 35 packets: " + str(t_E2E_35) + "ms")

End-to-End-Delay for 35 packets: 0.875ms


<H3> 3) Determine which packages will arrive at destination Z when Q sends 35 packages at once. Now, take into account the specified buffer capacities. <H3>

**Link1**<br>
*arrival-interval:* 0.1ms<br>
*transmission delay:* 0.5ms

| Zeitpunkt (ms) | Ereignis         | Buffer               | Übertragung | Ende        |
|:--------------:|:----------------:|:--------------------:|:-----------:|:-----------:|
| 0.0            | A1               |                      | P1          | 0.5         |
| 0.1            | A2               | A2                   |             |             |
| 0.2            | A3               | A3, A2               |             |             |
| 0.3            | A4               | A4, A3, A2           |             |             |
| 0.4            | A5               | A5, A4, A3, A2       |             |             |
| 0.5            | D1               | A5, A4, A3           | P2          | 1.0         |
|                | A6               | A6, A5, A4, A3       |             |             |
| 0.6            | A7 (dropped)     | A6, A5, A4, A3       |             |             |
| 0.7            | A8 (dropped)     | A6, A5, A4, A3       |             |             |
| 0.8            | A9 (dropped)     | A6, A5, A4, A3       |             |             |
| 0.9            | A10 (dropped)    | A6, A5, A4, A3       |             |             |
| 1.0            | D2               | A6, A5, A4           | P3          | 1.5         |
|                | A11              | A11, A6, A5, A4      |             |             |
| 1.1            | A12 (dropped)    | A11, A6, A5, A4      |             |             |
| 1.2            | A13 (dropped)    | A11, A6, A5, A4      |             |             |
| 1.3            | A14 (dropped)    | A11, A6, A5, A4      |             |             |
| 1.4            | A15 (dropped)    | A11, A6, A5, A4      |             |             |
| 1.5            | D3               | A11, A6, A5          | P4          | 2.0         |
|                | A16              | A16, A11, A6, A5     |             |             |


Starting from package 6, only every 5th package arrives
-> 11 out of 35 arrive: (A1-A6, A11, A16, A21, A26, A31)


**Link2**<br>
*arrival-interval:* 0.5ms<br>
*transmission delay:* 2.5ms

Due to the same ratio of packet arrival rate to data rate as in Link 1, one can expect the same result! Therefore, starting from packet 6, only every 5th packet arrives here as well! -> 7 out of 11: (A1-A6, A31).

**Link3**<br>
*arrival-interval:* 2.5ms<br>
*transmission delay:* 25ms

| Zeitpunkt (ms) | Ereignis         | Buffer               | Übertragung | Ende        |
|:--------------:|:----------------:|:--------------------:|:-----------:|:-----------:|
| 0.0            | A1               |                      | P1          | 25          |
| 2.5            | A2               | A2                   |             |             |
| 5.0            | A3               | A3, A2               |             |             |
| 7.5            | A4(dropped)      | A3, A2               |             |             |
| 10.0           | A5(dropped)      | A3, A2               |             |             |
| 12.5           | A6(dropped)      | A3                   |             |             |
| 15.0           | A31(dropped)     | A4, A3               |             |             |

In this case, only the first two packets arrive -> 2 out of 7: (A1, A2).